### DISCUSSION TITLE
E2E Tests are getting failed (Error: Ended WebDriver sessions gracefully after a SIGINT signal was received)

### DISCUSSION THREAD

**Discussion by vaibhav-gautam-by** [webdriverio/webdriverio#13515] [Community]
Hi,

When we (me and my team) are running multiple instances of e2e tests manually (or if they are running in multiple PR checks) we are facing this "Error: Ended WebDriver sessions gracefully after a SIGINT signal was received" and I have found that it only happens when there are multiple instances of e2e workflows running simultaneously (more than 10 instances at a time).

The version of wdio we are using is 8.10.2

Could you please help and tell what is the problem here. Also i am sharing the screenshots for you guys to have a look.

<img width="959" alt="image-2024-08-30-15-56-06-169" src="https://github.com/user-attachments/assets/4ebfdda1-ec51-4d24-a3cf-45be0bdbbd50">
<img width="956" alt="image-2024-08-30-15-55-34-117" src="https://github.com/user-attachments/assets/a4d6cb5b-0973-465b-b659-140a2d3f0911">

**Comment by christian-bromann** [comment]
I think this is a Github Action feature which cancels other pipelines automatically once you have 1 failing pipline.

**Reply by vaibhav-gautam-by** [reply]
Hi @christian-bromann , Thanks for the response.
 
We would like to understand the reason why this error comes up and how could we can prevent it.
 
We suspect comes when a large number of instances running simultaneously, So could you please confirm if there is any cap on number of e2e workflows that can be run at a time. If yes, how can we increase that number? Please share an example on how this can be configured.
 
FYI, Earlier the PR checks were passing even with this Error, (which is incorrect as gives the impression that our e2e ran without any issues). 

here you can find a screenshot as reference. We would expect that as there’s SIGINT error displayed the workflow stops.

![image (2)](https://github.com/user-attachments/assets/6b85cb8c-d41c-4454-87dc-dca2ea2e9dad)

And to prevent that from happening we have added as workaround a command tracing to catch the error and make the check fail. Which is not ideal as this hide the logs until the e2e run is completed.
 
here you can see example of the workaround mentioned:


 {
[# code ]

      - name: Running E2E tests
        id: run-e2e
        run: |
          set -x  # Enable command tracing for debugging

         [# some code executed here .....]

          set +x  # Disable command tracing
          if [[ "$OUTPUT" == *"Ended WebDriver sessions gracefully after a SIGINT signal was received"* ]]; then
            echo "Webdriver error message detected; failing step."
            exit 1
          else
            echo "No Webdriver error message found."
 }
 
just to recap, we would like to know
Is there a version available where SIGINT error is handled properly and make the check fail
what is the limit of instances that can run at the same time without get the SIGINT error
thanks in advance for your attention, I highly appreciate your help on this.

**Reply by christian-bromann** [reply]
> So could you please confirm if there is any cap on number of e2e workflows that can be run at a time.

You can run as many parallel tests as your CPU allows it with WebdriverIO, however it might not be smart to have a large number of tests on a single machine as this may introduce flakiness due to performance problems. Please take a look at https://webdriver.io/docs/sharding/ to learn about how you can shard tests using GitHub Actions and other CI/CD pipelines.

> here you can find a screenshot as reference.

I am not sure where the SIGINT signal comes from and why the tests were stopped. I don't think you can run 53 parallel sessions on GHA. Again, I recommend to look into sharding.

> Is there a version available where SIGINT error is handled properly and make the check fail

No, this is an invalid use case for WebdriverIO. When a SIGINT is received WebdriverIO ends the test run.

> what is the limit of instances that can run at the same time without get the SIGINT error

This depends on your environment. I recommend to not run more than 10 headless test on a single CI machine.
